crossbar/router/test/test_broker.py

```diff
@@ -511,6 +512,197 @@ class TestSession(ApplicationSession):
+                    router._broker.processUnsubscribe(self._service_session, subscription)
+                    subscriptions.append(obj.subscription)
+
+                def all_done():
+
+                    created = list(subscriptions)
+                    subscribes = list(subscriptions)
+                    unsubscribes = list(subscriptions)
+                    deletes = list(subscriptions)
+
+                    for args, argv in self._service_session._events:
+                        if args[0] == 'wamp.subscription.on_create':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_create: session id is incorrect!')
+                            test.assertTrue(args[2]['id'] in created, 'on_create: subscription id is incorrect!')
+                            created.remove(args[2]['id'])
+
+                        if args[0] == 'wamp.subscription.on_subscribe':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_subscribe: session id is incorrect!')
+                            test.assertTrue(args[2] in subscribes, 'on_subscribe: subscription id is incorrect!')
+                            subscribes.remove(args[2])
+
+                        if args[0] == 'wamp.subscription.on_unsubscribe':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_unsubscribe: session id is incorrect!')
+                            test.assertTrue(args[2] in unsubscribes, 'on_unsubscribe: subscription id is incorrect!')
+                            unsubscribes.remove(args[2])
+
+                        if args[0] == 'wamp.subscription.on_delete':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_delete: session id is incorrect!')
+                            test.assertTrue(args[2] in deletes, 'on_delete: subscription id is incorrect!')
+                            deletes.remove(args[2])
+
+                    test.assertEqual(len(created), 0, 'incorrect response sequence for on_create')
+                    test.assertEqual(len(subscribes), 0, 'incorrect response sequence for on_subscribe')
+                    test.assertEqual(len(unsubscribes), 0, 'incorrect response sequence for on_unsubscribe')
+                    test.assertEqual(len(deletes), 0, 'incorrect response sequence for on_delete')
+
+                reactor.callLater(1, all_done)
+                time.sleep(1)
+
+        session = TestSession(types.ComponentConfig(u'realm1'))
+        self.session_factory.add(session, authrole=u'trusted')
+
+    def test_subscribe_detach(self):
+        """
+        Make sure the wamp.* event sequence delivered as part of a subscribe/detach sequence
+        is correct and contains the correct session id and subscription id values.
+        """
+        router = self.router
+        test = self
+
+        class Session(mock.MagicMock):
+            """
```

Comment:
Does this need to subclass Mock? (i.e. are there other method-calls that we're expecting other than the ones in this fake Session?). If so, it might be a nicer pattern (and more explicit/obvious) to do something like:
```
class Session(object):
    some_method = Mock()
```

...or similar.

Reply:
Yeah, there's .. probably a lot of methods. I guess it just looked weird to me to subclass it, but amounts to about the same thing .. so, good like this :)